File: 0862-shortest-subarray-with-sum-at-least-k/0862-shortest-subarray-with-sum-at-least-k.java
import java.util.*;

class Solution {
    public int shortestSubarray(int[] nums, int k) {

        int n = nums.length;
        long[] prefix = new long[n + 1];

        for (int i = 0; i < n; i++) {
            prefix[i + 1] = prefix[i] + nums[i];
        }

        Deque<Integer> dq = new ArrayDeque<>();
        int ans = n + 1;

        for (int i = 0; i <= n; i++) {

            while (!dq.isEmpty() &&
                    prefix[i] - prefix[dq.peekFirst()] >= k) {

                ans = Math.min(ans, i - dq.pollFirst());
            }

            while (!dq.isEmpty() &&
                    prefix[i] <= prefix[dq.peekLast()]) {

                dq.pollLast();
            }

            dq.offerLast(i);
        }

        return ans == n + 1 ? -1 : ans;
    }
}

File: 0862-shortest-subarray-with-sum-at-least-k/0862-shortest-subarray-with-sum-at-least-k.py
from collections import deque

class Solution:
    def shortestSubarray(self, nums: List[int], k: int) -> int:
        n = len(nums)

        # Prefix sum
        prefix = [0] * (n + 1)
        for i in range(n):
            prefix[i + 1] = prefix[i] + nums[i]

        dq = deque()      # Stores indices of prefix sums
        ans = n + 1

        for i in range(n + 1):
            # Check if we have a valid subarray
            while dq and prefix[i] - prefix[dq[0]] >= k:
                ans = min(ans, i - dq.popleft())

            # Maintain increasing prefix sums
            while dq and prefix[i] <= prefix[dq[-1]]:
                dq.pop()

            dq.append(i)

        return ans if ans <= n else -1